contracts/base/ApproveAndCall.sol

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity =0.7.6;
+pragma abicoder v2;
+
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+import '../interfaces/IApproveAndCall.sol';
+import './ImmutableState.sol';
+
+abstract contract ApproveAndCall is IApproveAndCall, ImmutableState {
+    function tryApprove(address token, uint256 amount) private returns (bool) {
+        (bool success, bytes memory data) =
+            token.call(abi.encodeWithSelector(IERC20.approve.selector, positionManager, amount));
+        return success && (data.length == 0 || abi.decode(data, (bool)));
+    }
+
+    /// @dev Can be called off-chain to determine which of the more efficient functions below should be called
+    function approve(address token) external payable override returns (SuccessfulApproveType) {
```

Comment:
do we want a lens function in here? might be confusing, you can do it from any contract

also maybe a requiredApprovalType function would be useful that takes an amount, to handle the case where approval is already max/high enough

Reply:
yeah in my head this is basically intended to be used a lens function already, though we can't make it view because it's potentially state-changing. one interesting idea is that it could check its existing allowance as well? also, realized we don't need the approve 0 -> approve max functions because we can just stack multiple approves, so i'll work on making this more generic

Reply:
ok i realized that a single `approve` method that just gives the caller the ability to specify amount basically solves everything. you can already use this as a lens off-chain to see which calls fail/are required/etc., it allows 0 -> max approvals, basically everything we should need for most tokens